### PR TITLE
Josh biddick

### PR DESCRIPTION
**Pull request by sadsa**

### CODE REVIEW

**Comment by MatiasIac** [comment]
It could be really good to get rid of the array. It is a good solution, but it will force the developer to create an array with some specific properties. This could be confusing for the developer and will consume more memory.

But, as I said, it is a good approach.

**Comment by den-t** [comment]
Liking this one. 👍 
We need to either create a processing chain, linked list, array or parameter object to solve this problem.

A bit wordy initialisation here, potentially requiring documentation or jumping back in forth through the code. Love how structured this solution is, will be really easy to extend at any level.

**Comment by MatiasIac** [comment]
@sadsa is this PR ready to be merged?
